[fix][broker] Fix out-of-order issues with ConsistentHashingStickyKeyConsumerSelector

lhotari opened the pull request:
Fixes #23315
Fixes #23321

### Motivation

See #23315 and #23321. This PR fixes a bug in ConsistentHashingStickyKeyConsumerSelector where the selected consumer for a hash could change across 2 existing consumers when a consumer was removed or when a new consumer was added.
In addition it fixes the problem that ConsistentHashingStickyKeyConsumerSelector.getConsumerKeyHashRanges() returned invalid information. These are fixed together since they are in the same code area and fixing getConsumerKeyHashRanges is required for testing the consistent selection of a consumer when there are hash collisions.

### Modifications

In PR #8396 a solution was added to fix a problem where all hash ring points were assigned to the last arrived consumer when there were hash collisions, for example when the consumers had the same name:
https://github.com/apache/pulsar/blob/4f96146f13b136644a4eb0cf4ec36699e0431929/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConsistentHashingStickyKeyConsumerSelector.java#L127

That solution causes problems since it causes the bug #23315. A different solution is needed to distribute the consumers evenly when there are hash collisions.

The solution in this PR is to assign a consumer with a "consumer name index". When there are multiple consumers with the same name, the name index will be unique. Since any consumer name could be potentially used by consumers arriving later, it is necessary to track all consumer names. When all consumers have been removed for a specific consumer name, it shouldn't result in a memory leak. This is the reason why tracking reference counts is handled in the implementation. 

The implementation contains a solution where a consumer leaving and joining immediately will get assigned with the same "consumer name index". This improves stability since the hash assignment changes are minimized over time, although a better solution would be to avoid reusing the same consumer name in the first place.

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

lhotari commented:
I renamed this to "reduce" out-of-order issues since I'm not sure if the current model in this PR prevents the case that a consumer gets swapped when an unrelated consumer leaves.
I was able to get the behavior fairly consistent, but it's not going to prevent all cases. 

It seems that another hash ring layer is needed for the colliding consumers. That would prevent unnecessary changes. A hash ring inside a hash ring...

lhotari commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConsistentHashingStickyKeyConsumerSelector.java:
```diff
@@ -44,15 +46,105 @@ public class ConsistentHashingStickyKeyConsumerSelector implements StickyKeyCons
+        public void addConsumer(Consumer consumer, MutableInt selectedCounter) {
+            ConsumerEntry consumerEntry = new ConsumerEntry(consumer, selectedCounter);
+            consumers.add(consumerEntry);
+            if (selectedConsumerEntry == null || consumerEntry.compareTo(selectedConsumerEntry) < 0) {
+                // if the new consumer has a higher priority or lower selection count
+                // than the currently selected consumer, select the new consumer
+                changeSelectedConsumerEntry(consumerEntry);
+            }
```
I don't think that it's a breaking change. The previous behavior is already broken.

To fix the problem there needs to be a solution where the consumer is only changed when a new consumer is added or when the currently "selected" consumer is removed.

In the previous solution, all consumers for a colliding hash were active at the same time. The consumer was selected by the input hash on this line:
https://github.com/apache/pulsar/blob/4f96146f13b136644a4eb0cf4ec36699e0431929/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConsistentHashingStickyKeyConsumerSelector.java#L127
That was a wrong solution that was broken.

One small detail what I made in the sorting order was that I made it also consider the priority of the consumer. That's been ignored before, but I think it has been a mistake before. Since the previous behavior has been broken, we aren't introducing a breaking change when we fix it.

equanz replied:
After your comment, I debugged your code, and I noticed that my understanding was incorrect.
My understanding at the time was, "If a new consumer is added, the range will always move to a new one."

Now, I understand as follows. Is it correct?
"When the selected count was greater than new consumer, then the consumer's range is moved to a new one."

lhotari replied:
> After your comment, I debugged your code, and I noticed that my understanding was incorrect. My understanding at the time was, "If a new consumer is added, the range will always move to a new one."
> 
> Now, I understand as follows. Is it correct? "When the selected count was greater than new consumer, then the consumer's range is moved to a new one."

@equanz  In the previous solution `consumerList.get(hash % consumerList.size())` could result in a different consumer any time consumers are added or removed. Since the `hash` passed as the parameter is the hash calculated for the message key, it also depends on the input message. It could also result in the same consumer as before, so it doesn't mean that is always switched, however the probability increases when there are more consumers since hitting a different consumer has (n-1)/n chance.

lhotari commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConsistentHashingStickyKeyConsumerSelector.java:
```diff
@@ -44,15 +46,105 @@ public class ConsistentHashingStickyKeyConsumerSelector implements StickyKeyCons
+            return removed;
+        }
+
+        private void changeSelectedConsumerEntry(ConsumerEntry newSelectedConsumer) {
+            if (newSelectedConsumer == selectedConsumerEntry) {
+                return;
+            }
+            beforeChangingSelectedConsumerEntry();
+            selectedConsumerEntry = newSelectedConsumer;
+            afterChangingSelectedConsumerEntry();
+        }
+
+        private void beforeChangingSelectedConsumerEntry() {
+            if (selectedConsumerEntry != null) {
+                selectedConsumerEntry.consumerSelectionCounter.decrement();
+            }
+        }
+
+        private void afterChangingSelectedConsumerEntry() {
+            if (selectedConsumerEntry != null) {
+                selectedConsumerEntry.consumerSelectionCounter.increment();
+            }
+        }
+    }
```
There's no need to make this class thread safe. This class isn't exposed externally. All method calls are made under the read-write lock (`ReadWriteLock rwLock`). That's covering the thread safety.

lhotari commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConsistentHashingStickyKeyConsumerSelector.java:
```diff
@@ -44,15 +46,105 @@ public class ConsistentHashingStickyKeyConsumerSelector implements StickyKeyCons
+        }
+
+        public Consumer getSelectedConsumer() {
+            return selectedConsumerEntry != null ? selectedConsumerEntry.consumer() : null;
```
> This PR breaks the original behavior( the original behavior contains an issue ), which may cause a consumer not to receive any messages if users register consumers with the same name.

The hash ring will contain 100 entries and for each entry, the selected consumer will be evenly distributed. That's how it's already covered. Most of the complexity in this PR about addressing that issue that all registered consumers with the same name would get evenly distributed. 
This applies when there are less than 100 consumers. This can be configured by increasing `subscriptionKeySharedConsistentHashingReplicaPoints` if that's not sufficient. However, users should use unique consumer names if there are more than 100 consumers.

poorbarcode replied:
I think you have not answered my questions, such as follows

> The issue's occurring condition is that multiple consumers have the same hash( it always occurs when the consumer names are the same ), right?

> This PR breaks the original behavior( the original behavior contains an issue ), which may cause a consumer not to receive any messages if users register consumers with the same name.

poorbarcode replied:
> > This PR breaks the original behavior( the original behavior contains an issue ), which may cause a consumer not to receive any messages if users register consumers with the same name.

> No. I explained in the previous comment.

- The consumers that were calculated with the same `hash` will be pushed to the same `HashRingEntry`
- The selector will always get the same `HashRingEntry` for a certain `key`
- The `HashRingEntry` selected will always return the consumer who has the highest priority.
- The consumers who have a lower priority will never be selected

lhotari replied:
> - The consumers that were calculated with the same `hash` will be pushed to the same `HashRingEntry`
> - The selector will always get the same `HashRingEntry` for a certain `key`

That's the correct behavior and the intent of this PR. The previous behavior was not correct at all since it caused the bug.

There will be multiple hash ring entries (100 by default) and the colliding consumers are evenly selected across all of them. Did you already consider that aspect?


> - The `HashRingEntry` selected will always return the consumer who has the highest priority.
> - The consumers who have a lower priority will never be selected

It's true. The consumer priority was ignored before, but that has been a bug. Bugs should be fixed.

poorbarcode replied:
>> - The consumers that were calculated with the same `hash` will be pushed to the same `HashRingEntry`
>> - The selector will always get the same `HashRingEntry` for a certain `key`

> There will be multiple hash ring entries (100 by default) and the colliding consumers are evenly selected across all of them. Did you already consider that aspect?

Yes, but the multiple consumers who have the same name will be assigned the same 100 `HashRingEntry`s, right?

If yes, the consumers who have a lower priority will never be selected, they will receive nothing.

lhotari replied:
> Yes, but the multiple consumers who have the same name will be assigned the same 100 `HashRingEntry`s, right?

No. Most of the complexity in this PR is handle the case that multiple consumers with the same name will be evenly distributed as the assigned consumers for the 100 hash ring points.
Please take a look at the diff and the code comments for example for the `consumerSelectionCounters` field and how that gets used. The counter for each consumer keeps the state of the number of hash ring entries where a specific consumer is "selected" (assigned as the consumer). This counter is part of the ordering criterias for the ConsumerEntry. That's how it's determined that for consumers with the same priority and name, the lowest selection count consumer gets selected. Is the current code understandable or do you think that more comments should be added to explain this "magic"?

> If yes, the consumers who have a lower priority will never be selected, they will receive nothing. You broke the bahavior

Why would they set the priority if they don't want it to be used? I wouldn't call this in breaking behavior. We can also ask @merlimat for an opinion.

> I think consumer.priority is designed for Failover mode, so Key_shared discarded it.

It's not only for Failover mode. The javadoc explains that it's support for Shared mode too: https://github.com/apache/pulsar/blob/cefa72c7d18ab05d413e8737672f746629874f59/pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ConsumerBuilder.java#L508-L553 . It's a matter of defining what the behavior should be for Key_Shared subscriptions.

poorbarcode replied:
>> Yes, but the multiple consumers who have the same name will be assigned the same 100 `HashRingEntry`s, right?

> No. Most of the complexity in this PR is handle the case that multiple consumers with the same name will be evenly distributed as the assigned consumers for the 100 hash ring points.
Please take a look at the diff and the code comments for example for the `consumerSelectionCounters` field and how that gets used. The counter for each consumer keeps the state of the number of hash ring entries where a specific consumer is "selected" (assigned as the consumer).

Yes, I noticed `consumerSelectionCounters` before, but the values of this map are always `0`, which is meanless

lhotari replied:
> Yes, I noticed `consumerSelectionCounters` before, but the values of this map are always `0`, which is meanless

I don't see that happening. You could modify ConsistentHashingStickyKeyConsumerSelectorTest#shouldRemoveConsumersFromConsumerKeyHashRanges for example and make it use 100 points and a duplicate name. The counters are 1 in most cases which is a sign that this "magic" really works. There are some other Key_Shared tests that would fail if this didn't work.
<img width="1244" alt="image" src="https://github.com/user-attachments/assets/65dbb170-cbf6-45cf-973c-9b6bf584561a">
Do you see it now?

poorbarcode replied:
> Do you see it now?

Sure, I found it now, it was changed by `beforeChangingSelectedConsumerEntry` and `afterChangingSelectedConsumerEntry`.  

Since you used it to mark how many `points` it owned, your PR has another issue; I wrote a test to reproduce it.

```java
    @Test
    public void testConsumerReconnects() {
        final ConsistentHashingStickyKeyConsumerSelector selector = new ConsistentHashingStickyKeyConsumerSelector(100);
        final String consumerName = "consumer";
        final int numOfInitialConsumers = 50;
        final int validationPointCount = 200;
        final List<Integer> pointsToTest = pointsToTest(validationPointCount);
        List<Consumer> consumers = new ArrayList<>();
        for (int i = 0; i < numOfInitialConsumers; i++) {
            final Consumer consumer = createMockConsumer(consumerName, "index " + i, i);
            consumers.add(consumer);
            selector.addConsumer(consumer);
        }

        // Mark original results.
        List<Consumer> selectedConsumersBeforeRemove = new ArrayList<>();
        for (int i = 0; i < validationPointCount; i++) {
            int point = pointsToTest.get(i);
            selectedConsumersBeforeRemove.add(selector.select(point));
        }

        // Let consumers reconnect
        // Verify: the same range selected as "selectedConsumersBeforeRemove"
        for (Consumer c : consumers) {
            selector.removeConsumer(c);
        }
        for (int i = numOfInitialConsumers -1; i > -1 ; i--) {
            // Consumers reconnect.
            selector.addConsumer(consumers.get(i));
            for (int j = 0; j < validationPointCount; j++) {
                int point = pointsToTest.get(j);
                Consumer selected = selector.select(point);
                Consumer expected = selectedConsumersBeforeRemove.get(j);
                assertThat(selected.consumerId()).as("validationPoint %d, hash %d", j, point).isEqualTo(expected.consumerId());
            }
        }
    }

    private List<Integer> pointsToTest(int validationPointCount) {
        List<Integer> res = new ArrayList<>();
        int hashRangeSize = Integer.MAX_VALUE;
        final int increment = hashRangeSize / (validationPointCount + 1);
        for (int i = 0; i < validationPointCount; i++) {
            res.add(i * increment);
        }
        return res;
    }
```

lhotari replied:
> Since you used it to mark how many `points` it owned, your PR has another issue; I wrote a test to reproduce it.

@poorbarcode Thanks, this is a good way to handle the details. I'll check what could be supported in the solution. I provided a longer update in a separate comment https://github.com/apache/pulsar/pull/23327#issuecomment-2376288786. Without an additional key, it's not possible to solve this problem. The challenge is that the client IP address is mostly useless in many k8s deployments which don't forward the original client IP address.

I've taken an approach where I'm trying to add an additional key based on the arrival index of a specific consumer name. If the index is unused, it can be taken. This way a consumer that leaves and joins back would be possible to get the same assignment. However the code doesn't yet work properly.

lhotari replied:
@poorbarcode regarding the unit test `testConsumerReconnects`, does that pass with current master branch?

poorbarcode replied:
> @poorbarcode regarding the unit test testConsumerReconnects, does that pass with current master branch?

It can not pass, but what I want to say is you considered Pulsar should keep the key range for a consumer absolutely consistent, but you just fixed half of the issue. If you agree that you just want to solve half of the issue, I could provide a suggestion that is simpler.

lhotari replied:
> It can not pass, but what I want to say is you considered Pulsar should keep the key range for a consumer absolutely consistent, but you just fixed half of the issue. If you agree that you just want to solve half of the issue, I want to advise a new solution that is simpler than the current one.

This issue is primarily fixing the out-of-order issue which is a critical issue, #23315. It is caused with the fact that the current implementation will swap 2 existing consumers when other consumers connect or disconnect. 
Let's first make progress on addressing that issue in this PR. Ok?

lhotari replied:
> I tried the [test above](https://github.com/apache/pulsar/pull/23327#discussion_r1776320513) again, it can not pass even if I pulled the newest commits

Yes that's true, I haven't had a chance to dig in the details yet.

lhotari replied:
@poorbarcode The test is fixed in 80d64514fa35daa4d6e507c92b43a3a3313835bf . The problem in your test was that it was assuming that the assignments would be the same when there were different consumer counts. The current implementation will ensure that when consumers leave and join in the same order, they will get assigned with the same hash range assignments. This is a reasonable "best effort" approach when the application is assigning the same consumer name. As we have discussed, that is not a recommended approach in the first place for Key_Shared subscriptions. This PR is for fixing #23315 and #23321 with a reasonable approach.

codelipenghui reviewed (commented):
I'm thinking another simple solution to replace the `NavigableMap<Integer, List<Consumer>> hashRing;` to `NavigableMap<Integer, HashRangeAutoSplitStickyKeyConsumerSelector> hashRing;`. Since the `HashRangeAutoSplitStickyKeyConsumerSelector` used a fixed hash range and it can also distribute loads to multiple consumers. ConsistentHashingStickyKeyConsumerSelector + HashRangeAutoSplitStickyKeyConsumerSelector together will resolve the both load balance and hash collision issue.

@lhotari @poorbarcode Thanks for bring up the customer's priority discussion. I think the Key_Shared subscription missed the consumer's priority support before. The expected behavior for consumer's priority is if you have a higher priority consumer under the subscription, the consumer with lower priority should not receive message until the consumer with higher priority get disconnected. I would like to suggest to take this one out of this PR. This issue is not easy to resolve in the Consumer Selector since different key will select different consumers unless we remove all the low priority consumers out of the Consumer Selector.

lhotari commented:
> @lhotari @poorbarcode Thanks for bring up the customer's priority discussion. I think the Key_Shared subscription missed the consumer's priority support before. The expected behavior for consumer's priority is if you have a higher priority consumer under the subscription, the consumer with lower priority should not receive message until the consumer with higher priority get disconnected. I would like to suggest to take this one out of this PR. This issue is not easy to resolve in the Consumer Selector since different key will select different consumers unless we remove all the low priority consumers out of the Consumer Selector.

@codelipenghui  Valid point. I have removed the priority handling. We can come back to that after PIP-379 has been implemented since it's easy to solve there. The current PIP-379 implementation will also handle cases where hash ranges switch from one existing consumer to another existing consumer. That will be essential in addressing the priority handling. It's also possible to later add the logic that you are describing to the model that is in this PR.

lhotari commented:
> I'm thinking another simple solution to replace the `NavigableMap<Integer, List<Consumer>> hashRing;` to `NavigableMap<Integer, HashRangeAutoSplitStickyKeyConsumerSelector> hashRing;`. Since the `HashRangeAutoSplitStickyKeyConsumerSelector` used a fixed hash range and it can also distribute loads to multiple consumers. ConsistentHashingStickyKeyConsumerSelector + HashRangeAutoSplitStickyKeyConsumerSelector together will resolve the both load balance and hash collision issue.

@codelipenghui I was also thinking about that type of solution; however, it won't be effective. The reason for this is that each hash range in the actual hash ring is not evenly sized. Even if the embedded hash ring would perfectly distribute the duplicate consumers, it wouldn't solve the overall distibution.
That's also why the previous solution in this PR didn't work properly, even though the consumers were evenly assigned across all hash ranges when there were collisions. That's why any embedded hash ring or container solution work and I have reverted those changes from this PR and taken a different approach.

This would be a simple problem to solve with an additional key. I guess @poorbarcoded suggested using a client IP address or port. The client address would be a useful additional key, but in many k8s environments, the IP address is masked when the ingress solution doesn't sustain the original IP address. The IP address of a proxy isn't very useful.

I have refactored the approach in this PR to use a hash-based solution where the idea is that the additional key for the hash would be the arrival index of the consumer for a particular name. If a consumer leaves and reconnects immediately, it would retain the same ranges. If consumers leave and reconnect in different order, the result wouldn't be consistent from a single consumer's viewpoint. The consumer would have to have an unique name in that case.
In the current version there's no consistency for this type of case at all. `consumerList.get(hash % consumerList.size())` depends on the number of connected consumers and that will vary. It's not a "breaking change" to have a way to mitigate the duplicate consumer name behavior, even when there isn't full consistency between different phases.

The distribution is properly even now with the changes (I added a test case for proving that), but there are some remaining issues to fix since a test case derived from @poorbarcode's test case isn't yet passing locally.

lhotari commented:
One detail that I noticed in ee7a83e7945ca966e7b47fde589179469f4e71b9 is that by increasing the hash ring points from 100 to 200, it will improve the even distribution across the consumers a lot.

lhotari commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConsistentHashingStickyKeyConsumerSelector.java:
```diff
@@ -53,55 +56,127 @@ public ConsistentHashingStickyKeyConsumerSelector(int numberOfPoints) {
-                });
+                int consumerNameIndex = consumerNameIndexTracker.addHashRingReference(consumerIdentityWrapper);
+                int hash = calculateHashForConsumerAndIndex(consumer, consumerNameIndex, i);
+                ConsumerIdentityWrapper removed = hashRing.put(hash, consumerIdentityWrapper);
```
> In my understanding is that the current approach "attempts to create different keys even if the consumerName is the same. However, in this approach, when hashes collide, it is replaced by the one added later." Is this understanding correct?

@equanz Yes, that's correct. However the hash collision won't happen due to using the same consumer name since there will be a unique key for each consumer with the same consumer name. 
When the previous is replaced, it won't cause existing consumers to switch for a specific hash. 

Any sort of implementation where there's a sub list or data structure for colliding consumers will result in uneven distribution since each hash range has a different size. That's why all entries must be in a single hash ring for the Key_shared Pulsar use case. I have made extensive experimentation and research to back this. If anyone is interested in more details, I can share them. I shared some details my previous comment, https://github.com/apache/pulsar/pull/23327#issuecomment-2376288786.

equanz replied:
> the hash collision won't happen due to using the same consumer name since there will be a unique key consumerNameIndex for each consumer with the same consumer name.

I see. Thank you.

I think this PR no longer guarantees against coincidental hash collisions referred to in the https://github.com/apache/pulsar/pull/8396.
However, it depends on the definition of what should be guaranteed(e.g. always consistent, almost consistent and always distributed).
What do you think?

lhotari replied:
> I think this PR no longer guarantees against coincidental hash collisions referred to in the #8396.
> However, it depends on the definition of what should be guaranteed(e.g. always consistent, almost consistent and always distributed).
> What do you think?

The previous problem with collisions with the consumer name was that all consumers had collisions on all hash ring points (100 by default). That is solved in this PR by ensuring that when consumers have the same consumer name, there's an additional key, `consumerNameIndex`. That's how there won't be collisions on all hash ring points. Individual collisions aren't a problem since there will be 100 hash ring points by default. If there's a single collision, the other hash ring points will ensure that there won't be any problems in practice. I don't think that there's any drawbacks on consistency because of the 100 hash ring points. If one consumer has 1 point less due to a collision, the impact is statistically in the range of a few percentages. 

Even without any changes coming from this PR, the distribution is already not accurately distributed with 100 hash ring points. Currently the error margin is roughly in the range of 20-30% with 100 hash ring points. With 200 points, the accuracy improves and variation reduces, (IIRC about 15% in my experimentation). Also from this view point, having hash collisions on single hash ring points doesn't impact the accuracy significantly.

equanz replied:
Thnaks for your description.
I probably already know your point, but it is not my concern. My concern is the definition of ConsistentHashing's guarantee of consistency. If we change the definition of consistency, at least we should add some comments about it.

By the way, in my opinion, the above(changing the definition of consistency) is not a problem in Key_Shared since Key_Shared does not guarantee that the hash range is consistent regardless of the connection order. In fact, the HashRangeAutoSplitStickyKeyConsumerSelector depends on the connection order.
And the above is not a problem in ConsistentHashingStickyKeyConsumerSelector. In https://github.com/apache/pulsar/pull/6791 the consistent hashing is introduced for load balancing reasons. It is not intended to be consistent regardless of connection order.

I think this PR is a great work.

codelipenghui replied:
> The previous problem with collisions with the consumer name was that all consumers had collisions on all hash ring points (100 by default). That is solved in this PR by ensuring that when consumers have the same consumer name, there's an additional key, consumerNameIndex. That's how there won't be collisions on all hash ring points. Individual collisions aren't a problem since there will be 100 hash ring points by default. If there's a single collision, the other hash ring points will ensure that there won't be any problems in practice. I don't think that there's any drawbacks on consistency because of the 100 hash ring points. If one consumer has 1 point less due to a collision, the impact is statistically in the range of a few percentages.

@lhotari But user can also set the points to 1 for some reason since it's configurable. Or even a smaller value, not one. If we can accept this case, the hash collisions from different consumer names should also be fine, just replace the old point, like the old way without https://github.com/apache/pulsar/pull/8396

My point is that we can make the change more simple

1. Change the consumer key, have consumer_name, consumer_id, address (it still work with non-proxy case)
2. Just have a rehash method for the conflict point

Something like this:
```java
for (int i = 0; i < numberOfPoints; i++) {
    int hash = calculateHashForConsumerAndIndex(consumer, i);
    Consumer existConsumer = hashRing.putIfAbsent(hash, consumer);
    while (existConsumer != null) {
        hash = rehash(hash);
        existConsumer = hashRing.putIfAbsent(hash, consumer);
    }
}
```

Both replace or rehash has it's pros and cons, and the cons for both of them are not a big problem.

- Replace, we are leveraging the points and message keys for the load distribution. One replace might also cause the existing consumer cannot get any messages.
- Rehash, consumer will not always added to the same point when hash collisions happened. But it should be fine, it's only a tiny part of all the consumers.

So, I prefer a less change, straightforward solution.

@lhotari

codelipenghui replied:
But I don't want to say it's a blocker for this PR. @lhotari If you still want to continue with your solution in this PR, I'm fine with it. I can continue to review the details.

lhotari replied:
> My point is that we can make the change more simple
> 
> 1. Change the consumer key, have consumer_name, consumer_id, address (it still work with non-proxy case)

Adding consumer_id will break the current intended behavior that when a consumer disconnects and reconnects immediately, that it gets assigned with the same hash ranges as before the disconnection happened. When using unique consumer names, the behavior change would be a significant regression. There's no way to determine beforehand whether a consumer name will be unique or not. That's why the logic would need to be used in all cases.

Adding address isn't useful at all for most users of Pulsar. The socket ports aren't useful since a new socket would be used in most cases after reconnecting after a connection is terminated and re-established. It wouldn't be an effective solution.

> 2. Just have a rehash method for the conflict point

Makes sense for handling the conflict.

> So, I prefer a less change, straightforward solution.

Please point to exact details of what is "much change" if you say that there's a lot of changes in this PR. It will help understand your point. Perhaps there's a way to improve this PR to match your criteria. Let's stay open for that. This PR is addressing a bug fix with a change that minimizes the risk of regressions. I believe that it's a common goal for us.

poorbarcode replied:
> My point is that we can make the change more simple
> Change the consumer key, have consumer_name, consumer_id, address (it still work with non-proxy case)
Just have a rehash method for the conflict point
> Something like this: ...

Agree with @codelipenghui

lhotari replied:
> Agree with @codelipenghui, we can choose a simpler implementation if we just want to fix the out-of-order issue.

@poorbarcode this has been addressed in the most recent changes. If you disagree about the simplicity, please provide specific feedback which part of the code in this PR is too complex in your mind.

lhotari replied:
> And the major part is the rehash solution is more straightforward. Readers can read it and understand it in 1 min without understand why it's safe for replacing the point,

@codelipenghui I have refactored the PR to follow this rule. Please take another look. 
My expectation is that you could point to exact details in the code of this PR if there are concerns that the code is too complex. I hope that is a reasonable ask from my side.

codelipenghui commented:
@lhotari How about just adding a rehash method for the consumer will put to the same position in the Hash Ring? The benefits for this solution:

- Less memory overhead (we don't create to create list, maps for each point)
- No major changes
- Easy to read and understand

As you mentioned increasing the hash ring points from 100 to 200". It should be more common way for users to get a better distribution to increase the ring points, but for the existing solution or new solution in this PR will use much more heap memory that user expected because we will create list or map for each point. If they have 100 consumers (not a very large case), we will have 20k lists or maps.

lhotari commented:
> @lhotari How about just adding a rehash method for the consumer will put to the same position in the Hash Ring? The benefits for this solution:

@codelipenghui There wouldn't be any consistency in that type of solution. In the solution in this PR, the consumer will get assigned the same hash ranges if it disconnects and comes immediately back. 

> * Less memory overhead (we don't create to create list, maps for each point)

There's not much memory overhead in the solution in this PR. I don't see how this could be a problem in this PR.

> * No major changes

This PR doesn't contain a "major change"

> * Easy to read and understand

Please be more specific of this feedback. What isn't easy to read and understand in this PR and I could try to improve it.

lhotari commented:
> As you mentioned increasing the hash ring points from 100 to 200". It should be more common way for users to get a better distribution to increase the ring points, but for the existing solution or new solution in this PR will use much more heap memory that user expected because we will create list or map for each point. If they have 100 consumers (not a very large case), we will have 20k lists or maps.

This PR doesn't increase inaccuracy in any way. The inaccuracy is an existing property of the hash ring solution and not related to the hash collisions.

lhotari commented:
I have refactored the code in this PR. I hope it now appears how simple the code is and there isn't a need for alternative solutions.  Please take another look. /cc @codelipenghui  @merlimat @equanz @poorbarcode

lhotari commented:
Hash ranges are very rare with the current hash range size of `2^31-1` (`Integer.MAX_VALUE`).
In PIP-379, the hash range size is reduced to `2^16-1` (`65535`). In that case, hash range conflicts are common.
However, due to the large amount of hash ring points, there's no significant difference in distributions due to the conflicts.

equanz reviewed (approved):
LGTM
